code/traffic-register/traffic-register.js

```diff
@@ -551,3 +557,114 @@ $('#view_657 tbody input[type=checkbox]:checked').each(function() {
+  which relies on $<field #>$ to be replaced with the value of
+  a field from a newly-created canvas_regulation record. */
+  "ANGLEPARK":
+    "$421$ $391$ - on -  $394$ from $405$ $406$ of $403$ to $408$, $399$, $414$ side(s).",
+};
+
+/* Some fields return HTML in field_<field #> key, so we need to grab from the raw field
+which contains an array of objects. */
+var fieldsToUseRawData = {
+  391: "391_raw",
+  394: "394_raw",
+  403: "403_raw",
+  408: "408_raw",
+  1313: "1313_raw",
+};
+
+function makeAutomatedRegulationText(record){
+  var regulationTypeField = "field_1176";
+
+  // Get the pattern by regulation type
+  var regulationType = record[regulationTypeField];
+  var pattern =
+    automatedRegulationTextPatternsByType[regulationType];
+
+  // We should only generate the automated regulation text if there is a pattern defined.
+  if(pattern){
+    // Gather the fields that need to be replaced with their values
+    // Use match all to get matches and the captured strings too
+    // Ex. [["$391$", "391"], ...]
+    var regex = /\$(.*?)\$/g;
+    var matchesIterable = pattern.matchAll(regex);
+
+    // Collect matches and values in an array of objects
+    // For example, [{match: "$391$", value: "value-from-Knack-record"}, ...]
+    var matches = [];
+    for (var matchArray of matchesIterable) {
+      var matchObject = {
+        match: matchArray[0],
+      };
+
+      var fieldNumber = matchArray[1];
+      // Create the field key to get the value from the record
+      var fieldKey;
+      if (fieldNumber in fieldsToUseRawData) {
+        fieldKey = "field_" + fieldsToUseRawData[fieldNumber];
+      } else {
+        fieldKey = "field_" + fieldNumber;
+      }
+
+      // Add the value to the match object
+      if (fieldNumber in fieldsToUseRawData) {
+        // Get the value out of the array of objects in raw data if there is one
+        var value = record[fieldKey][0] ? record[fieldKey][0].identifier : "" ;
+        matchObject["value"] = value;
+      } else {
+        var value = record[fieldKey];
+        matchObject["value"] = value;
+      }
+
+      matches.push(matchObject);
+    }
+
+    // Go through the map and replace the pattern with the value
+    var automatedRegulationText = pattern;
+    matches.forEach(function (matchObject) {
+      var match = matchObject.match;
+      var value = matchObject.value;
+      automatedRegulationText = automatedRegulationText.replace(match, value);
+    });
+
+    // Async request to populated field_658 with the automated text
+    var headers = {
+      "X-Knack-Application-ID": appSpecifics.knackAppId,
+      Authorization: Knack.getUserToken(),
+      "content-type": "application/json",
+    };
+
+    var putUrl =
+      "https://api.knack.com/v1/pages/" +
+      appSpecifics.apiSceneIdForRegulationText +
+      "/views/" +
+      appSpecifics.apiViewIdForRegulationText +
+      "/records/";
+    var recordId = record.id;
+
+    $.ajax({
```

Comment:
Curious if any knows if we have a pattern for retries? Looks like we routinely get away without them on production apps but wanted to ask.

Reply:
I don't think we have one. I went to look at [this code](https://github.com/cityofaustin/atd-knack/blob/020dd9235cb33a20ebf64ee79222d5c71a4fc25d/code/finance-puchasing/finance-purchasing.js#L627) that iirc you wrote, but we're not re-trying.

I'm a little bit wary of relying on this post-form submit record update, because the user may not know when the request fails, or have a way to fix it, right? Without blowing out the scope of this, would it be feasible to get the reg text assembled before the form submit? either by live-updating the tex in the form, or intercepting the submit button click?

Reply:
Yeah, I thought about this, but I thought there wasn't a path to do that. I'll check it out because I'm could totally be misremembering and would also prefer that path if possible. 🙏